logging/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+    fn from(v: CommunicationEvent) -> CommsEvent { CommsEvent::Communication(v) }
+}
+
+impl From<SerializationEvent> for CommsEvent {
+    fn from(v: SerializationEvent) -> CommsEvent { CommsEvent::Serialization(v) }
+}
+
+impl From<OperatesEvent> for Event {
+    fn from(v: OperatesEvent) -> Event { Event::Operates(v) }
+}
+
+impl From<ChannelsEvent> for Event {
+    fn from(v: ChannelsEvent) -> Event { Event::Channels(v) }
+}
+
+impl From<ProgressEvent> for Event {
+    fn from(v: ProgressEvent) -> Event { Event::Progress(v) }
+}
+
+impl From<PushProgressEvent> for Event {
+    fn from(v: PushProgressEvent) -> Event { Event::PushProgress(v) }
+}
+
+impl From<MessagesEvent> for Event {
+    fn from(v: MessagesEvent) -> Event { Event::Messages(v) }
+}
+
+impl From<ScheduleEvent> for Event {
+    fn from(v: ScheduleEvent) -> Event { Event::Schedule(v) }
+}
+
+impl From<ApplicationEvent> for Event {
+    fn from(v: ApplicationEvent) -> Event { Event::Application(v) }
+}
+
+impl From<GuardedMessageEvent> for Event {
+    fn from(v: GuardedMessageEvent) -> Event { Event::GuardedMessage(v) }
+}
+
+impl From<GuardedProgressEvent> for Event {
+    fn from(v: GuardedProgressEvent) -> Event { Event::GuardedProgress(v) }
+}
+
+impl From<CommChannelsEvent> for Event {
+    fn from(v: CommChannelsEvent) -> Event { Event::CommChannels(v) }
+}
+
+impl From<InputEvent> for Event {
+    fn from(v: InputEvent) -> Event { Event::Input(v) }
+}
+
+const BUFFERING_LOGGER_CAPACITY: usize = 1024;
+
+pub enum LoggerBatch<'a, S: Clone+'a, L: Clone+'a> {
+    Logs(&'a Vec<(u64, S, L)>),
+    End,
+}
+
+enum BufferingLoggerInternal<S: Clone, L: Clone> {
+    Active {
+        setup: S,
+        buffer: RefCell<Vec<(u64, S, L)>>,
+        pushers: RefCell<Box<Fn(LoggerBatch<S, L>)->()>>,
+    },
+    Inactive
+}
+
+pub struct BufferingLogger<S: Clone, L: Clone> {
+    internal: BufferingLoggerInternal<S, L>,
+}
+
+impl<S: Clone, L: Clone> BufferingLogger<S, L> {
+    pub fn new(setup: S, pushers: Box<Fn(LoggerBatch<S, L>)->()>) -> Self {
+        BufferingLogger {
+            internal: BufferingLoggerInternal::Active {
+                setup: setup,
+                buffer: RefCell::new(Vec::with_capacity(BUFFERING_LOGGER_CAPACITY)),
+                pushers: RefCell::new(pushers),
+            },
+        }
+    }
+
+    pub fn new_inactive() -> Rc<BufferingLogger<EventsSetup, Event>> {
+        Rc::new(BufferingLogger {
+            internal: BufferingLoggerInternal::Inactive,
+        })
+    }
+
+    pub fn log(&self, l: L) {
+        match self.internal {
+            BufferingLoggerInternal::Active { ref setup, ref buffer, ref pushers } => {
+                let ts = get_precise_time_ns();
+                let mut buf = buffer.borrow_mut();
+                buf.push((ts, setup.clone(), l));
+                if buf.len() >= BUFFERING_LOGGER_CAPACITY {
+                    (*pushers.borrow_mut())(LoggerBatch::Logs(&buf));
+                    buf.clear();
```

Comment:
What do you think about the `Push` and `Pull` idioms from the communication crate? Rather than a function that takes a `&Vec<_>` which is then cleared afterwards, you use a function that takes an `Option<&mut Vec<_>>`, which allows the recipient to swap the backing memory if they like (e.g. if they just want to stash the `Vec` in a linked list or something), or just `drain` out of it, as appropriate.